### PR TITLE
Minor: improve ParquetOpener docs

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/4028

## Rationale for this change

While working with @itsjunetime on https://github.com/apache/datafusion/pull/12135 we spent a while reviewing the code in `ParquetOpener` and I realized that the documentation could be improved

## What changes are included in this PR?

Improve the documentation

## Are these changes tested?
By CI

## Are there any user-facing changes?
No functional changes, just docs

### CODE REVIEW

File: datafusion/core/src/datasource/physical_plan/parquet/opener.rs [hunk 1]
```diff
@@ -41,21 +41,41 @@ use std::sync::Arc;
 
 /// Implements [`FileOpener`] for a parquet file
 pub(super) struct ParquetOpener {
+    /// Execution partition index
     pub partition_index: usize,
+    /// Column indexes in `table_schema` needed by the query
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/opener.rs:46]: This in particular was unclear when @itsjunetime  and I read the code (were the indexes in terms of file_schema or table_schema?) and was the impetus for this PR

**Review by appletreeisyellow — APPROVED** [review]
Thank you @alamb! The descriptions are helpful, especially those for bool values!

**Review by comphead — APPROVED** [review]
Nice, WDYT about enabling the `missing_docs` lint so every pub will force forced to be documented ?

**Comment by comphead** [comment]
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e486601506c2493123d2b4d9cca24727
here is the link to playground, the lint will require crate documentation and public entities

**Comment by alamb** [comment]
> Nice, WDYT about enabling the `missing_docs` lint so every pub will force forced to be documented ?

I personally think it is a good idea, though I predict that if we do that we'll still end up with functions that are like "ParquetOpener: opens parquet" any other similar documentation that isn't particularly helpful.

But it might encourage enough that is better than no documentation 👍

**Comment by alamb** [comment]
Thanks for the review @appletreeisyellow  and @comphead

**Comment by comphead** [comment]
> I personally think it is a good idea, though I predict that if we do that we'll still end up with functions that are like "ParquetOpener: opens parquet" any other similar documentation that isn't particularly helpful.
> 
> But it might encourage enough that is better than no documentation 👍

I'll try to play on some small crate first, enabling the documentation and putting `#[allow(missing_docs)]` whenever it complains, so later it would be easier to find all undocumented methods and make contributions

**Pull request merged**